### PR TITLE
first pull request

### PR DESCRIPTION
**Pull request by danyeboah**
First week done

### CODE REVIEW

File: app/assets/javascripts/users.js.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (brandoncc) [inline app/assets/javascripts/users.js.coffee]: We recommend you create your models and controllers manually in this course. That will do two things for you:
1. You will learn a lot more about how models and controllers work.
2. You will not have a ton of extra files in your project like these.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class SessionsController < ApplicationController
+  def new
+    redirect_to videos_path if logged_in?
+  end
+
+  def create
+    user = User.find_by(email: params[:email])
+
+    if user && user.authenticate(params[:password])
+      if(params[:remember_me]) == "1"
```
Comment (brandoncc) [inline app/controllers/sessions_controller.rb:10]: Nice job adding remember me functionality.

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        cookies.permanent.signed[:auth_token] = user.auth_token
+      else
+        cookies.signed[:auth_token] = user.auth_token
+      end
+
+      flash["notice"] = "You have been successfully logged in"
```
Comment (brandoncc) [inline app/controllers/sessions_controller.rb]: Bootstrap 3 (which is used by this project) uses "success" and "danger" instead of "notice" and "alert".

Comment (brandoncc) [inline app/controllers/sessions_controller.rb]: oops I mean "error", not "alert"

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+class UsersController < ApplicationController
+  
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.create(user_params)
```
Comment (brandoncc) [inline app/controllers/users_controller.rb]: Create saves already. You want to do `User.new`, then the next line handles the attempted save.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+class User < ActiveRecord::Base
+  include Slug
+  sluggable_column :lastname
+  before_save :generate_slug
+
+  validates_presence_of :firstname
```
Comment (brandoncc) [inline app/models/user.rb]: Your columns should be "snake_case" not smushedtogether.

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+  before_save do 
+    generate_token(:auth_token)
+  end
+
+  # generate random token for login
+  def generate_token(column)
+    begin
+      self[column] = SecureRandom.urlsafe_base64
+    end while User.exists?(column == self[column])
```
Comment (brandoncc) [inline app/models/user.rb]: This is not doing what you think it is. You are doing this:

``` ruby
User.exists?(column == self[column])
```

Which is equivalent to:

``` ruby
User.exists?(:auth_token == 'some_token')
```

which is equivalent to:

``` ruby
User.exists?(false)
```

This is what you want to do:

``` ruby
User.exists?(column => self[column])
```

You need to provide a hash which has the column name and value you are looking for.

http://apidock.com/rails/ActiveRecord/Base/exists%3F/class

File: spec/models/category_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    it 'should return 6 videos ordered by recency(created) where there are more than 6 videos' do
+      The_League = Video.create(title: "The League", description: "Bros", category: @comedy, created_at: 1.day.ago)
+      Futurama = Video.create(title: "Futurama", description: "Cool Story", category: @comedy, created_at: 3.days.ago)
+      South_Park = Video.create(title: "South Park", description: "Too funny",category: @comedy, created_at: 8.days.ago)
+      The_Future = Video.create(title: "The Future", description: "Cool Baby",category: @comedy, created_at: 2.days.ago)
+      Scandal = Video.create(title: "Scandal", description: "nice", category: @comedy, created_at: 5.days.ago)
+      Lost = Video.create(title: "Lost", description: "sweet", category: @comedy, created_at: 6.days.ago)
+
+      expect(@comedy.recent_videos(6)).to eq([The_League,The_Future,Futurama,Scandal,Lost,South_Park])
```
Comment (brandoncc) [inline spec/models/category_spec.rb]: Your variables should be snake_case. You are creating constants instead of local variables.

**Comment by danyeboah** [comment]
Made all the changes suggested.

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      redirect_to videos_path
+    else 
+      flash["error"] = "Username or password is invalid"
+      redirect_to login_path
+    end
+  end
+
+  def destroy
+    cookies.delete(:auth_token)
+    flash["notice"] = "You have successfully logged out"
```
Comment (brandoncc) [inline app/controllers/sessions_controller.rb]: Still using notice instead of success.

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
+class UsersController < ApplicationController
+  
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)     
+    if @user.save
+      flash["success"] = "Your account has been created" \
```
Comment (brandoncc) [inline app/controllers/users_controller.rb]: No need for the backslash.

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+      redirect_to login_path
+    else
+      render :new
+    end
+  end
+
+  def user_params
+    params.require(:user).permit!
```
Comment (brandoncc) [inline app/controllers/users_controller.rb]: This is very dangerous. If someone adds admin=1 to their post, they are now an admin. You definitely want to use strong parameters.

File: app/models/video.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  validates_presence_of :title, :description
+  belongs_to :category, foreign_key: :category_id
+
+  sluggable_column :title
+
+  before_save :generate_slug
+
+
+  def self.search_by_title(title)
+    where("lower(title) LIKE ?", "%#{title.downcase}%").order("title ASC  ")
```
Comment (brandoncc) [inline app/models/video.rb]: You can also use `ILIKE` for case insensitive searches, just so you know.

File: app/views/sessions/new.html.haml [hunk 12]
```diff
@@ -0,0 +1,13 @@
+%section.sign_in.container
+  .row
+    .col-sm-10.col-sm-offset-1
+      %header
+        %h1.lead Sign in
+        %hr.hr1
+
+        =bootstrap_form_tag url: '/login', method: 'post', layout: :horizontal, label_col: "col-sm-2", control_col: "col-sm-3" do |f|
```
Comment (brandoncc) [inline app/views/sessions/new.html.haml]: You should use named paths instead of hard coding them.

File: spec/models/video_spec.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    end
+
+    it "returns an empty array where no match" do
+      result = Video.search_by_title('boss')
+      expect(result).to eq([])
+    end
+
+    it "returns an array with a single item if there is perfect match" do
+      result = Video.search_by_title('South Park')    
+      expect(result).to eq([South_Park])
```
Comment (brandoncc) [inline spec/models/video_spec.rb]: Do your specs pass? You are expecting capitalized (constant) variables here and in spec/models/category_spec.rb. They shouldn't pass.

**Comment by danyeboah** [comment]
changes made

**Pull request merged**